Interestingly, this fails our test:

```
        # Leaving DST, clocks are rolled back
>       assert datetime(2018, 10, 28, 2, 30, 0, tzinfo=tz, fold=0) == find(
            datetime(2018, 10, 28, 2, 5, 0, tzinfo=tz, fold=0), 2, 30, 0
        )
E       AssertionError: assert datetime.datetime(2018, 10, 28, 2, 30, tzinfo=backports.zoneinfo.ZoneInfo(key='Europe/Vienna')) == datetime.datetime(2018, 10, 29, 2, 30, tzinfo=backports.zoneinfo.ZoneInfo(key='Europe/Vienna'))
```

--- a/homeassistant/util/dt.py
+++ b/homeassistant/util/dt.py
@@ -357,11 +357,11 @@ def find_next_time_expression_time(
 def datetime_exists(dattim: dt.datetime) -> bool:
     """Check if a datetime exists."""
     assert dattim.tzinfo is not None
-    tz = dattim.tzinfo
+    original_tzinfo = dattim.tzinfo
     # Check if we can round trip to UTC
-    return dattim.replace(tzinfo=None) == dattim.astimezone(UTC).astimezone(tz).replace(
-        tzinfo=None
-    )
+    return dattim.replace(tzinfo=None) == dattim.astimezone(UTC).astimezone(
+        original_tzinfo
+    ).replace(tzinfo=None)
 
 
 def datetime_ambiguous(dattim: dt.datetime) -> bool:

--- a/homeassistant/util/dt.py
+++ b/homeassistant/util/dt.py
@@ -355,23 +355,21 @@ def find_next_time_expression_time(
 
 
 def datetime_exists(dattim: dt.datetime) -> bool:
-    """Check if a datetime exists.
-
-    From https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html#detecting-ambiguous-and-imaginary-times
-    """
-    # There are no non-existent times in UTC, and comparisons between
-    # aware time zones always compare absolute times; if a datetime is
-    # not equal to the same datetime represented in UTC, it is imaginary.
-    return dattim.astimezone(dt.timezone.utc) == dattim
+    """Check if a datetime exists."""
+    assert dattim.tzinfo is not None
+    tz = dattim.tzinfo
+    # Check if we can round trip to UTC
+    return dattim.replace(tzinfo=None) == dattim.astimezone(UTC).astimezone(tz).replace(
+        tzinfo=None
+    )
 
 
 def datetime_ambiguous(dattim: dt.datetime) -> bool:
-    """Check whether a datetime is ambiguous.
-
-    From https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html#detecting-ambiguous-and-imaginary-times
-    """
-    # If a datetime exists and its UTC offset changes in response to
-    # changing `fold`, it is ambiguous in the zone specified.
-    return datetime_exists(dattim) and (
-        dattim.replace(fold=not dattim.fold).utcoffset() != dattim.utcoffset()
+    """Check whether a datetime is ambiguous."""
+    assert dattim.tzinfo is not None
+    non_folded = dattim.replace(fold=0)
+    folded = dattim.replace(fold=1)
+    return not (
+        non_folded.utcoffset() == folded.utcoffset()
+        and non_folded.dst() == folded.dst()
     )

--- a/homeassistant/util/dt.py
+++ b/homeassistant/util/dt.py
@@ -318,7 +318,7 @@ def find_next_time_expression_time(
     if result.tzinfo in (None, UTC):
         return result
 
-    if datetime_ambiguous(result):
+    if _datetime_ambiguous(result):
         # This happens when we're leaving daylight saving time and local
         # clocks are rolled back. In this case, we want to trigger
         # on both the DST and non-DST time. So when "now" is in the DST
@@ -327,7 +327,7 @@ def find_next_time_expression_time(
         if result.fold != fold:
             result = result.replace(fold=fold)
 
-    if not datetime_exists(result):
+    if not _datetime_exists(result):
         # This happens when we're entering daylight saving time and local
         # clocks are rolled forward, thus there are local times that do
         # not exist. In this case, we want to trigger on the next time
@@ -344,17 +344,17 @@ def find_next_time_expression_time(
     # For example: if triggering on 2:30 and now is 28.10.2018 2:30 (in DST)
     # we should trigger next on 28.10.2018 2:30 (out of DST), but our
     # algorithm above would produce 29.10.2018 2:30 (out of DST)
-    if datetime_ambiguous(now):
+    if _datetime_ambiguous(now):
         check_result = find_next_time_expression_time(
             now + _dst_offset_diff(now), seconds, minutes, hours
         )
-        if datetime_ambiguous(check_result):
+        if _datetime_ambiguous(check_result):
             return check_result
 
     return result
 
 
-def datetime_exists(dattim: dt.datetime) -> bool:
+def _datetime_exists(dattim: dt.datetime) -> bool:
     """Check if a datetime exists."""
     assert dattim.tzinfo is not None
     original_tzinfo = dattim.tzinfo
@@ -364,7 +364,7 @@ def datetime_exists(dattim: dt.datetime) -> bool:
     ).replace(tzinfo=None)
 
 
-def datetime_ambiguous(dattim: dt.datetime) -> bool:
+def _datetime_ambiguous(dattim: dt.datetime) -> bool:
     """Check whether a datetime is ambiguous."""
     assert dattim.tzinfo is not None
     non_folded = dattim.replace(fold=0)

